File: content/development/golang-first-impression/call_by_xxx.java
class CallByXXX {
	public CallByXXX() {
		this.a = 0;
	}
	public int a;

	public static void callByReference(CallByXXX obj) {
		obj.a += 1;
		System.out.println(String.format("in function : %d", obj.a));
	}

	public static void main(String [] args) {
		CallByXXX obj = new CallByXXX();

		System.out.println(String.format("before call by reference : %d", obj.a));
		callByReference(obj);
		System.out.println(String.format("after call by reference : %d", obj.a));
	}
}

File: content/development/golang-first-impression/call_by_xxx.go
package main

import "fmt"

type Sample struct {
	a int
}

func callByValue(s Sample) {
	s.a += 1
	fmt.Println("in function : ", s.a)
}

func callByReference(s *Sample) {
	s.a += 1
	fmt.Println("in function : ", s.a)
}

func main() {
	s := Sample{0}

	fmt.Println("before call by value : ", s.a)
	callByValue(s)
	fmt.Println("after call by value : ", s.a)

	fmt.Println("before call by reference : ", s.a)
	callByReference(&s)
	fmt.Println("after call by reference : ", s.a)
}
